internal/config/config.go

```diff
@@ -0,0 +1,225 @@
+package config
+
+import (
+	"context"
+	"flag"
+	"log"
+	"os"
+	"strconv"
+	"time"
+
+	"go.uber.org/zap/zapcore"
+
+	"github.com/OleG2e/collector/pkg/logging"
+
+	"github.com/caarlos0/env/v11"
+	"go.uber.org/zap"
+)
+
+const (
+	defaultReportIntervalSeconds = 10
+	defaultPollIntervalSeconds   = 2
+	defaultStoreIntervalSeconds  = 300
+)
+
+type (
+	AgentConfig struct {
+		LogLevel       string `env:"LOG_LEVEL"`
+		Address        string `env:"ADDRESS"`
+		ReportInterval int    `env:"REPORT_INTERVAL"`
+		PollInterval   int    `env:"POLL_INTERVAL"`
+	}
+	ServerConfig struct {
+		LogLevel        string `env:"LOG_LEVEL"`
+		Address         string `env:"ADDRESS"`
+		StoreInterval   int    `env:"STORE_INTERVAL"`
+		FileStoragePath string `env:"FILE_STORAGE_PATH"`
+		Restore         bool   `env:"RESTORE"`
+	}
+)
+
+func NewAgentConfig(ctx context.Context, l *logging.ZapLogger) *AgentConfig {
+	var (
+		addr     string
+		logLevel string
+		ri       int
+		pi       int
+	)
+
+	c := AgentConfig{}
+
+	err := env.Parse(&c)
+
+	l.DebugCtx(ctx, "init agent env config",
+		zap.String("LOG_LEVEL", os.Getenv("LOG_LEVEL")),
+		zap.String("ADDRESS", os.Getenv("ADDRESS")),
+		zap.String("REPORT_INTERVAL", os.Getenv("REPORT_INTERVAL")),
+		zap.String("POLL_INTERVAL", os.Getenv("POLL_INTERVAL")),
+	)
+
+	if err != nil {
+		l.PanicCtx(ctx, "parse agent config env error", zap.Error(err))
+	}
+
+	flag.StringVar(&logLevel, "log_level", "info", "log level")
+	flag.StringVar(&addr, "a", "localhost:8080", "server address")
+	flag.IntVar(&ri, "r", defaultReportIntervalSeconds, "report interval")
+	flag.IntVar(&pi, "p", defaultPollIntervalSeconds, "poll interval")
+
+	flag.Parse()
+
+	l.DebugCtx(ctx, "init agent flags config",
+		zap.String("LOG_LEVEL", logLevel),
+		zap.String("ADDRESS", addr),
+		zap.Int("REPORT_INTERVAL", ri),
+		zap.Int("POLL_INTERVAL", pi),
+	)
+
+	if c.LogLevel == "" {
+		c.LogLevel = logLevel
+	}
+	if c.Address == "" {
+		c.Address = addr
+	}
+	if c.ReportInterval == 0 {
+		c.ReportInterval = ri
+	}
+	if c.PollInterval == 0 {
+		c.PollInterval = pi
+	}
+
+	l.DebugCtx(ctx, "final agent params",
+		zap.String("ADDRESS", c.Address),
+		zap.Int("REPORT_INTERVAL", c.ReportInterval),
+		zap.Int("POLL_INTERVAL", c.PollInterval),
+		zap.String("LOG_LEVEL", c.LogLevel),
+	)
+
+	return &c
+}
+
+func NewServerConfig(ctx context.Context, l *logging.ZapLogger) *ServerConfig {
```

Comment:
```suggestion
func NewServerConfig(ctx context.Context, l *logging.ZapLogger) (*ServerConfig, error) {
```
И вместо дергания везде паники `return err`. А наверху уже можно либо log.Fatal, либо панику.